akka-projection-jdbc/src/test/scala/akka/projection/jdbc/JdbcProjectionSpec.scala

```diff
@@ -895,50 +897,64 @@ class JdbcProjectionSpec
         offsetOpt shouldBe empty
       }
 
-      val failingProjection =
+      def buildProjection(failPredicate: Long => Boolean = _ => false) =
         JdbcProjection
           .atLeastOnce(
             projectionId,
             sourceProvider = sourceProvider(system, entityId),
             jdbcSessionFactory,
-            handler = () => new ConcatHandler(_ == 4))
+            handler = () => new ConcatHandler(failPredicate))
           .withSaveOffset(2, 1.minute)
 
+      val failingProjection = buildProjection(_ == 4)
+
       withClue("check - offset is empty") {
         val offsetOpt = offsetStore.readOffset[Long](projectionId).futureValue
         offsetOpt shouldBe empty
       }
 
+      withClue("check: projection handles 3 elements") {
+        projectionTestKit.runWithTestSink(failingProjection) { sinkProbe =>
+          sinkProbe.request(3)
+          eventually {
+            val concatStr = findById(entityId)
+            concatStr.text shouldBe "abc|def|ghi"
+          }
+        }
+      }
+
+      withClue(s"check: last seen offset is 2L") {
+        eventually {
+          val offset = offsetStore.readOffset[Long](projectionId).futureValue.value
+          offset shouldBe 2L
+        }
+      }
+
       withClue("check: projection failed with stream failure") {
         projectionTestKit.runWithTestSink(failingProjection) { sinkProbe =>
-          sinkProbe.request(1000)
+          sinkProbe.request(2) // processes item 3 again and fails when processing 4
           eventuallyExpectError(sinkProbe).getMessage should startWith(concatHandlerFail4Msg)
         }
       }
       withClue("check: projection is consumed up to third") {
         val concatStr = findById(entityId)
-        concatStr.text shouldBe "abc|def|ghi"
+        concatStr.text shouldBe "abc|def|ghi|ghi" // note item 3 is processed twice
       }
       withClue(s"check: last seen offset is 2L") {
-        val offset = offsetStore.readOffset[Long](projectionId).futureValue.value
-        offset shouldBe 2L
+        eventually {
+          val offset = offsetStore.readOffset[Long](projectionId).futureValue.value
+          offset shouldBe 2L
+        }
```

Comment:
... start the projection again and cause the failure.

Observe the offset didn't advance.

Hmmm, I should probably advance the projection more carefully (demanding one item each time and awaiting for the eventual state to be persisted).